parser.cpp

```diff
@@ -67,7 +70,10 @@ Parser::Parser(
     if (p != std::string::npos)
         basename_ = std::string(
             basename_.begin(), basename_.begin() + static_cast<ptrdiff_t>(p));
-    lex_ = new Lexer(f);
+
+    Preprocessor cpp(defines_);
+    std::string processed = cpp.process(f, basename_);
+    lex_ = new Lexer(processed);
```

Comment:
The challenge with preprocessor is:
- How do we make the Lexer error messages refer to original filename
- How do we make the error locations refer to the location in the original file.

Reply:
I think the second one is just a matter of finding the right parameters to cpp. For the first one, I'll look into it.

Reply:
Great.
FYI, The intel repo issue that I created for incorrect line number has been open more more than 1.5 years: https://github.com/intel/linux-sgx/issues/319

So if you can find a solution that would be really cool.

Reply:
Ok, it was less simple than I thought.

Also the Windows preprocessor adds an extra newline to the beginning of the file which would also mess up line numbers. I don't think the existing tooling has a mechanism or producing the output we want.

One option would be to use a preprocessor library. Boost appears to have a preprocessor library/tool called "wave" that could be a better option.